### ISSUE TITLE
Initial Setup of neurotrader

### ISSUE THREAD

**Issue by julianbro** [julianbro/aistempre#1]
create a Python package neurotrader that trains and serves a multi‑input, multi‑horizon, probabilistic Transformer for financial time‑series. It must (1) predict next price (or log‑return), (2) classify short‑term trend, (3) classify long‑term trend, and (4) output calibrated probabilities and prediction intervals. Provide full project scaffolding, CLI tools, configs, tests, and docs.

Technology choices

Python 3.11

PyTorch + PyTorch Lightning (training loops), torchmetrics

Hydra (YAML configs), Pydantic (settings), Typer (CLI)

Pandas/Polars (ETL), TA‑Lib or ta (technical indicators), NumPy

Optuna and Ray Tune PBT (population‑based training); DEAP or Nevergrad for evolutionary search

Scikit‑learn (scalers, calibration), statsmodels (optional)

Backtesting: vectorbt (preferred) or backtesting.py

Logging: MLflow or WandB (optional, behind a flag)

Testing: pytest

Packaging: pyproject.toml (hatchling or poetry)

Project structure

neurotrader/
  pyproject.toml
  README.md
  LICENSE
  .env.example
  configs/
    data.yaml
    model.yaml
    train.yaml
    loss.yaml
    features.yaml
    tune.yaml
  src/neurotrader/
    __init__.py
    cli.py                         # Typer CLI entrypoints
    utils/
      seed.py, logging.py, time.py, io.py
    data/
      sources.py                   # abstract DataSource interface
      ccxt_source.py               # optional live API via ccxt
      csv_source.py                # local CSV/Parquet loader
      resample.py                  # align 1m, 15m, 4h, 1d, 1w
      splitter.py                  # PurgedWalkForwardSplitter
      dataset.py                   # MultiResolutionTorchDataset
      datamodule.py                # LightningDataModule
    features/
      registry.py                  # FeatureRegistry (plug‑in)
      technical.py                 # RSI, MACD, EMA, ATR, etc.
      price.py                     # returns, log returns, VWAP
      volatility.py                # RV, GARCH proxy, z‑vol
      microstructure.py            # spread, imbalance (if avail)
      calendar.py                  # day/time/holiday encodings
      crossasset.py                # optional cross‑asset inputs
    labels/
      targets.py                   # next‑return, short/long trend
      thresholds.py                # epsilon bands for flat class
    models/
      ms_transformer.py            # Multi‑Scale Transformer encoder
      heads.py                     # regression dist head + cls heads
      blocks.py                    # attention blocks, patching
      embeddings.py                # timeframe/instrument embeddings
    losses/
      factory.py                   # LossFactory (MSE, Huber, Quantile, GaussianNLL, StudentTNLL)
      calibration.py               # temperature scaling, isotonic
      conformal.py                 # conformal intervals for reg.
    training/
      train_loop.py                # LightningModule with multitask loss
      callbacks.py                 # early stop, checkpoint, EMA, clipping
      metrics.py                   # RMSE, sMAPE, DA, F1, AUROC, ECE, Brier
    tuning/
      optuna_runner.py
      ray_pbt.py
      evolutionary.py              # DEAP/Nevergrad to optimize hparams & thresholds
    inference/
      predictor.py                 # batch/stream inference
      postprocess.py               # price reconstruction, p(correct)
    backtest/
      simple_strategy.py           # thresholded long/short demo
      evaluate.py                  # Sharpe, Sortino, MDD, turnover
  scripts/
    train.py, predict.py, backtest.py, tune.py, calibrate.py, export_onnx.py
  tests/
    test_losses.py, test_splitter.py, test_labels.py, test_model_shapes.py

Core functionality & acceptance criteria

Inputs / multi‑resolution ETL

Accept configurable timeframes: 1m, 15m, 4h, 1d, 1w (defaults). Base clock in UTC, strict left‑aligned resampling without leakage.

Merge OHLCV across timeframes into aligned panels. Fill missing by forward‑fill within session bounds; no future info allowed.

Allow extra input “metrics” via FeatureRegistry with YAML config (features.yaml). Each feature has:

name, module:function, params, required lookback, output dtype/shape, and per‑timeframe enable flags.

RobustScaler/QuantileTransformer fit only on training folds; apply on val/test.

Targets (labels)

Next‑price prediction via log‑return (log(close_t+Δ / close_t)) for a configurable next‑step horizon (default Δ=1 base‑bar).

Short‑term trend: multi‑class (UP/DOWN/FLAT) over a configurable horizon (e.g., 30m or 2h) with epsilon band for FLAT.

Long‑term trend: multi‑class (UP/DOWN/FLAT) over a longer horizon (e.g., 1d or 1w).

Class weights optional; label smoothing optional.

Model (Transformer)

MultiScalePriceTransformer:

For each timeframe, patch the series (e.g., patch_len & stride) → linear projection → timeframe embedding + relative positional encodings.

N_tf encoder blocks per timeframe (local attention + feed‑forward).

Cross‑timeframe fusion via cross‑attention or gated pooling into a shared latent.

Optional instrument and regime embeddings.

Heads:

Distributional regression head outputs parameters for Normal or Student‑t (μ, σ[, ν]) over the next log‑return. Support quantile head (e.g., q ∈ {0.1, 0.5, 0.9}).

Short‑trend head (softmax) and Long‑trend head (softmax).

Dropout, layernorm, pre‑norm, residuals; gradient checkpointing flag.

Losses (easy to change)

Implement LossFactory with Hydra‐configurable selection:

Regression: mse, mae, huber(delta=1.0), quantile([0.1,0.5,0.9]), gaussian_nll, student_t_nll.

Classification: cross_entropy, focal(gamma=2), optional class weights.

Multi‑task loss = w_reg * L_reg + w_short * L_short + w_long * L_long from loss.yaml.

Probability and “probability of correct output”

For classification heads, output calibrated probabilities using temperature scaling (fit on val set) and report ECE and Brier.

For regression:

If Gaussian/Student‑t, compute prediction interval and p(|error| ≤ ε) for configurable ε (e.g., ε = 10 bps). Use the predictive distribution CDF.

Provide conformal prediction alternative for distribution‑free coverage; report empirical coverage.

Expose in inference: p_correct_class = max(calibrated_probs) and p_within_eps for regression.

Training & evaluation

LightningModule with AdamW, cosine decay w/ warmup, grad clipping, optional EMA of weights.

Purged walk‑forward CV splitter (no leakage across folds).

Metrics: RMSE, MAE, sMAPE for regression; accuracy, F1, AUROC, Directional Accuracy, MCC for trends; Sharpe, Sortino, Max Drawdown from a simple thresholded strategy in backtest/.

EarlyStopping and ModelCheckpoint by composite val objective (weighted sum or Optuna study direction).

Tuning (evolutionary / PBT)

Optuna search space: d_model, n_layers, n_heads, dropout, patch_len, stride, horizons, ε thresholds, loss weights, LR, weight decay.

Ray Tune PBT: periodically perturb LR, dropout, loss weights; select on composite score = αval_RMSE^-1 + βval_F1 + γ*Sharpe.

Evolutionary (DEAP/Nevergrad): optimize discrete/continuous hyperparameters and strategy thresholds (entry/exit, stop/take) using validation Sharpe and penalty for turnover.

Inference & serving

neurotrader-predict loads a checkpoint + scaler + calibrator; ingests a live/latest window; outputs:

next‑price estimate (reconstruct from log‑return), short/long trend probabilities, prediction intervals, and p(correct) as described.

Export to ONNX; example FastAPI server (optional).

Configs (example)

model.yaml: {variant: base|medium|large, d_model, n_heads, n_layers_tf, n_layers_fusion, ff_mult, dropout, patch_len, stride, context_len, horizons: {next:1, short:"30m", long:"1w"}}

features.yaml: enable/disable features per timeframe with params.

loss.yaml: {reg: gaussian_nll, cls: cross_entropy, weights: {reg:0.5, short:0.25, long:0.25}, epsilon_bps:10}

train.yaml: batch_size, max_epochs, precision, accelerator, devices, seed, gradient_clip_val, lr, wd, warmup_steps.

tune.yaml: search spaces and objective weights.

data.yaml: data source (csv/ccxt), symbol(s), timeframes=[1m,15m,4h,1d,1w], start/end, cache dirs.

CLI examples (Typer)

neurotrader-train --config-name train.yaml

neurotrader-tune --backend optuna --n-trials 100

neurotrader-tune --backend ray-pbt --time-budget-hours 6

neurotrader-calibrate --checkpoint path.ckpt

neurotrader-predict --checkpoint path.ckpt --latest-csv data/latest.csv

neurotrader-backtest --predictions preds.parquet --cash 100000

Tests

Unit tests for splitter (purge works), LossFactory swaps, label horizons, model forward shapes, calibration coverage.

Docs

README with data prep, config examples, and caveats (no financial advice; beware overfitting/leakage; walk‑forward evaluation).

Deliverables: fully runnable package with example data (toy CSV), default configs, and at least one working training run on CPU/GPU.

📐 Recommended model sizes & parameters

Choose a preset in model.yaml (variant: base|medium|large). These are sensible defaults for single‑symbol training; scale down if VRAM limited or context is long.

Base (good starting point)

d_model: 256, n_heads: 8, ff_mult: 4

n_layers_tf: 2 per timeframe encoder; n_layers_fusion: 2

dropout: 0.1, patch_len: 16, stride: 8

context_len: 512 (per timeframe after patching)

horizons: {next: 1 bar, short: "30m", long: "1w"}

Loss weights: {reg: 0.5, short: 0.25, long: 0.25}

Optim: AdamW (LR=2e-4, weight_decay=0.05), cosine decay, warmup 2% of steps

Medium (more capacity)

d_model: 512, n_heads: 8, ff_mult: 4

n_layers_tf: 3, n_layers_fusion: 3

dropout: 0.1–0.2, patch_len: 32, stride: 16

context_len: 1024

Same horizons; enable gradient checkpointing

Large (when you have data & GPU)

d_model: 768, n_heads: 12, ff_mult: 4

n_layers_tf: 4, n_layers_fusion: 4

dropout: 0.1–0.2, patch_len: 32–64, stride: 16–32

context_len: 1536–2048

Consider local attention (sliding‑window) to control memory

Tip: If the base timeframe is 1m, keep context_len * stride / base_tf ≤ a few days to avoid stale patterns.

📥 Inputs to include (beyond OHLCV)

Enable these through features.yaml (toggle per timeframe):

Returns/price transforms: log returns (1–N lags), cumulative return over horizon, z‑scored price, rolling mean/reversion gap.

Trend/volatility: EMA/SMA (various windows), MACD, RSI, Stoch, CCI, ADX, ATR, Bollinger bandwidth, Donchian channel position.

Volume & flow: OBV, VWAP, money flow index, volume z‑score, volume delta.

Volatility proxies: realized volatility (sqrt of summed intrabar returns²), Parkinson, Garman–Klass.

Microstructure (if available): bid‑ask spread, order‑book imbalance, trade sign imbalance.

Cross‑asset: relevant indices/futures/FX (e.g., DXY, VIX, ES), same features applied.

Calendar/regime: hour‑of‑day, day‑of‑week, month, session flags; regime state via rolling k‑means or HMM (embedded as categorical).

Risk filters: recent drawdown, max excursion, realized Sharpe over rolling window.

Keep a feature registry so adding/removing is one line in YAML.

🎯 Targets & “probability of correct output”

Next price: predict next log‑return; reconstruct price as
price_pred = last_price * exp(mu_pred) (for Gaussian/Student‑t heads).

Short/long trends: 3‑class (UP/DOWN/FLAT). Define UP if cumulative return over horizon > +epsilon_bps, DOWN if < -epsilon_bps, else FLAT. Configure ε in basis points.

Probability of correctness

Classification: after temperature scaling, p_correct = max(probs). Track ECE and Brier to ensure calibration.

Regression: if Normal/Student‑t:
p_within_eps = CDF( +ε / σ̂ ) − CDF( −ε / σ̂ ).
With conformal intervals, use the empirical coverage of the (1−α) band as the reliability measure.

🔁 Loss swapping (examples for loss.yaml)
reg: gaussian_nll        # options: mse|mae|huber|quantile|gaussian_nll|student_t_nll
reg_params: { }
cls: cross_entropy       # or focal with gamma:
cls_params: { }
weights: { reg: 0.5, short: 0.25, long: 0.25 }
epsilon_bps: 10

You should be able to switch losses without touching code by changing this YAML. The LossFactory wires it up inside the LightningModule.

🧪 Tuning: evolutionary & PBT

Population‑Based Training (Ray Tune): vary LR, dropout, loss weights, patch_len, horizons; exploit/explore every K epochs using composite score (val loss + Sharpe).

Evolutionary (DEAP/Nevergrad): genome includes:

{d_model, n_heads, n_layers_tf, n_layers_fusion, dropout, patch_len, stride, epsilon_bps, thresholds_for_backtest}

Fitness: 0.5 * (-val_RMSE_norm) + 0.25 * DA_short + 0.25 * Sharpe_val − λ * turnover

Optuna: TPE or CMA‑ES sampler for finer search; prune on no improvement.

🧷 Guardrails (critical for real data)

No leakage: fit scalers on train only; compute features using only past bars; purge split boundaries.

Time alignment: strict UTC; resample higher timeframes from base data; verify OHLC aggregation rules; do not forward‑fill across sessions.

Walk‑forward evaluation: report OOS metrics per fold; never tune on test.

Calibration: always recalibrate after any retraining/tuning.

Risk disclaimer: outputs are for research; not financial advice.
